File: js/multiAddress.js
const { createHash } = require("crypto");
const { keccak256 } = require("js-sha3");

function toChecksumAddress(hexBody) {
  const hash = keccak256(hexBody);
  let result = "0x";
  for (let i = 0; i < hexBody.length; i++) {
    result += parseInt(hash[i], 16) >= 8
      ? hexBody[i].toUpperCase()
      : hexBody[i];
  }
  return result;
}

const OKX_PREFIX = "XKO";

/**
 * Normalize an address-like string into a 0x-prefixed EVM address.
 *
 * Accepts:
 *  - "0x" prefixed hex string (40 hex chars after prefix)
 *  - plain hex string (40 chars)
 *  - "XKO" prefixed hex string (40 chars after prefix)
 *
 * Returns:
 *  - 0x-prefixed lowercase 40-hex-character address
 */
function toEvmAddress(input) {
  if (typeof input !== "string") {
    throw new Error("Address must be a string");
  }

  let s = input.trim();
  s = s.toLowerCase();

  const prefixLower = OKX_PREFIX.toLowerCase();
  if (s.startsWith("0x")) {
    s = s.slice(2);
  } else if (s.startsWith(prefixLower)) {
    s = s.slice(3);
  }

  if (s.length !== 40) {
    throw new Error(`Invalid address length: expected 40 hex chars, got ${s.length}`);
  }

  if (!/^[0-9a-fA-F]{40}$/.test(s)) {
    throw new Error("Invalid hex characters in address");
  }
  return toChecksumAddress(s);
}

/**
 * Convert an EVM address (0x-prefixed or not) into an XKO-prefixed address.
 *
 * Accepts:
 *  - "0x" prefixed hex string (40 hex chars)
 *  - plain hex string (40 chars)
 *
 * Returns:
 *  - "XKO" + 40 hex chars (lowercase)
 */
function fromEvmAddress(input) {
  if (typeof input !== "string") {
    throw new Error("Address must be a string");
  }

  let s = input.trim();
  s = s.toLowerCase();
  if (s.startsWith("0x")) {
    s = s.slice(2);
  }

  if (s.length !== 40) {
    throw new Error(`Invalid address length: expected 40 hex chars, got ${s.length}`);
  }

  if (!/^[0-9a-fA-F]{40}$/.test(s)) {
    throw new Error("Invalid hex characters in address");
  }
  const checksummed = toChecksumAddress(s);
  const without0x = checksummed.startsWith("0x") ? checksummed.slice(2) : checksummed;
  return OKX_PREFIX + without0x;
}


module.exports = { toEvmAddress, fromEvmAddress };

File: go/multiAddress.go
package address

import (
	"encoding/hex"
	"errors"
	"fmt"
	"regexp"
	"strings"

	"golang.org/x/crypto/sha3"
)

const okxPrefix = "XKO"

var hexBody = regexp.MustCompile(`^[0-9a-f]{40}$`)

// normalizeBody trims, lowers, strips acceptable prefixes, and validates length/hex.
func normalizeBody(input string, allowXko bool) (string, error) {
	if input == "" {
		return "", errors.New("address must be a string")
	}

	s := strings.TrimSpace(strings.ToLower(input))

	switch {
	case strings.HasPrefix(s, "0x"):
		s = s[2:]
	case allowXko && strings.HasPrefix(s, strings.ToLower(okxPrefix)):
		s = s[3:]
	}

	if len(s) != 40 {
		return "", fmt.Errorf("invalid address length: expected 40 hex chars, got %d", len(s))
	}
	if !hexBody.MatchString(s) {
		return "", errors.New("invalid hex characters in address")
	}
	return s, nil
}

func toChecksumAddress(body string) (string, error) {
	bytes, err := hex.DecodeString(body)
	if err != nil {
		return "", err
	}

	hasher := sha3.NewLegacyKeccak256()
	hasher.Write([]byte(strings.ToLower(body)))
	hash := hasher.Sum(nil)
	hashHex := hex.EncodeToString(hash)

	result := make([]byte, 2+len(body))
	copy(result, "0x")

	for i := 0; i < len(body); i++ {
		ch := body[i]
		if ch >= '0' && ch <= '9' {
			result[2+i] = ch
			continue
		}
		if hashHex[i] >= '8' {
			result[2+i] = strings.ToUpper(string(ch))[0]
		} else {
			result[2+i] = ch
		}
	}

	_ = bytes // keeps linter happy if you remove extra logic later
	return string(result), nil
}

func ToEvmAddress(input string) (string, error) {
	body, err := normalizeBody(input, true)
	if err != nil {
		return "", err
	}
	return toChecksumAddress(body)
}

func FromEvmAddress(input string) (string, error) {
	body, err := normalizeBody(input, false)
	if err != nil {
		return "", err
	}
	checksum, err := toChecksumAddress(body)
	if err != nil {
		return "", err
	}
	return okxPrefix + checksum[2:], nil
}